spark2.4/src/main/scala/KafkaReader.scala

```diff
@@ -102,11 +113,17 @@ object KafkaReader {
       reader.option("kafka.client.id", s"$group-$clientId")  // probably should use uuid instead of user input TODO
     }
     
-    val values = reader
-      .load.select(from_json(col("value") cast "string", schema, Map( "timestampFormat" -> "yyyy/MM/dd HH:mm:ss" )) as "data", col("timestamp") cast "long" as "TM_EXT_KAFKA")
-      .select("data.*", "TM_EXT_KAFKA")
-      .withColumn("TM_EXT_KAFKA", col("TM_EXT_KAFKA") * 1000)
-      .withColumn("TM_SSDS", unix_timestamp * 1000 )
+    val values = if( useFlowMarkers ) {
+      reader
+        .load.select(from_json(col("value") cast "string", schema, Map( "timestampFormat" -> "yyyy/MM/dd HH:mm:ss" )) as "data", col("timestamp") cast "long" as "TM_EXT_KAFKA")
+        .select("data.*", "TM_EXT_KAFKA")
```

Comment:
Why not to "join" this `select` with the above `select`? Why two?

Reply:
I’m open to suggestions, but I believe the code is in line with the examples shown at:
https://databricks.com/blog/2017/04/26/processing-data-in-apache-kafka-with-structured-streaming-in-apache-spark-2-2.html
https://sparkbyexamples.com/spark/spark-streaming-with-kafka/
https://stackoverflow.com/questions/42506801/how-to-use-from-json-with-kafka-connect-0-10-and-spark-structured-streaming
https://jaceklaskowski.github.io/mastering-spark-sql-book/spark-sql-functions-collection/

I also just tried some alternatives like the following, and they all resulted in the exception that follows.
`.load.select(from_json(col("value") cast "string", schema, Map( "timestampFormat" -> "yyyy/MM/dd HH:mm:ss" )), col("timestamp") cast "long" as "TM_EXT_KAFKA")`
`.withColumn("TM_EXT_KAFKA", col("TM_EXT_KAFKA") * 1000)`
`.withColumn("TM_SSDS", unix_timestamp * 1000 )`

`Exception in thread "main" java.lang.IllegalArgumentException: Can't get JDBC type for struct<ID:bigint,PAYLOAD:string,SRC_SERVER:string,SRC_THREAD:bigint,TM_GENERATED:bigint>`

The data needs to be pulled out of the struct.

Trying explode resulted in a different exception.

`.load.select(explode(from_json(col("value") cast "string", schema, Map( "timestampFormat" -> "yyyy/MM/dd HH:mm:ss" ))), col("timestamp") cast "long" as "TM_EXT_KAFKA")`
`.withColumn("TM_EXT_KAFKA", col("TM_EXT_KAFKA") * 1000)`
`.withColumn("TM_SSDS", unix_timestamp * 1000 )`

`Exception in thread "main" org.apache.spark.sql.AnalysisException: cannot resolve 'explode(jsontostructs(CAST(`value` AS STRING)))' due to data type mismatch: input to function explode should be array or map type, not struct`